spark-kernel-dsv2/src/main/java/io/delta/spark/dsv2/utils/ExpressionUtils.java

```diff
@@ -0,0 +1,261 @@
+   * @return Optional containing the Delta Kernel Literal, or empty if the value is null or of an
+   *     unsupported type
+   */
+  @VisibleForTesting
+  static Optional<Literal> convertValueToKernelLiteral(Object value) {
+    // TODO: convert null to NULL literal.
+    if (value == null) return Optional.empty();
+
+    if (value instanceof Boolean) {
+      Boolean b = (Boolean) value;
+      return Optional.of(Literal.ofBoolean(b));
+    }
+    if (value instanceof Byte) {
+      Byte b = (Byte) value;
+      return Optional.of(Literal.ofByte(b));
+    }
+    if (value instanceof Short) {
+      Short s = (Short) value;
+      return Optional.of(Literal.ofShort(s));
+    }
+    if (value instanceof Integer) {
+      Integer i = (Integer) value;
+      return Optional.of(Literal.ofInt(i));
+    }
+    if (value instanceof Long) {
+      Long l = (Long) value;
+      return Optional.of(Literal.ofLong(l));
+    }
+    if (value instanceof Float) {
+      Float f = (Float) value;
+      return Optional.of(Literal.ofFloat(f));
+    }
+    if (value instanceof Double) {
+      Double d = (Double) value;
+      return Optional.of(Literal.ofDouble(d));
+    }
+    if (value instanceof java.math.BigDecimal) {
+      // Preserve precision and scale from the original BigDecimal
+      java.math.BigDecimal bd = (java.math.BigDecimal) value;
+      return Optional.of(Literal.ofDecimal(bd, bd.precision(), bd.scale()));
+    }
+    if (value instanceof String) {
+      String s = (String) value;
+      return Optional.of(Literal.ofString(s));
+    }
+    if (value instanceof Character) {
+      Character c = (Character) value;
+      return Optional.of(Literal.ofString(c.toString()));
+    }
+    if (value instanceof char[]) {
+      // Convert char array to string
+      char[] charArray = (char[]) value;
+      return Optional.of(Literal.ofString(String.valueOf(charArray)));
+    }
+    if (value instanceof byte[]) {
+      byte[] arr = (byte[]) value;
+      return Optional.of(Literal.ofBinary(arr));
+    }
+    if (value instanceof Date) {
+      // Convert java.sql.Date to days since epoch
+      Date date = (Date) value;
+      return Optional.of(Literal.ofDate(InternalUtils.daysSinceEpoch(date)));
+    }
+    if (value instanceof Timestamp) {
+      // Convert java.sql.Timestamp to microseconds since epoch
+      Timestamp timestamp = (Timestamp) value;
+      return Optional.of(Literal.ofTimestamp(InternalUtils.microsSinceEpoch(timestamp)));
+    }
+    if (value instanceof LocalDate) {
+      // Convert java.time.LocalDate to days since epoch
+      LocalDate localDate = (LocalDate) value;
+      return Optional.of(Literal.ofDate((int) localDate.toEpochDay()));
+    }
+    if (value instanceof Instant) {
+      // Convert java.time.Instant to microseconds since epoch
+      Instant instant = (Instant) value;
+      long micros = ChronoUnit.MICROS.between(Instant.EPOCH, instant);
+      return Optional.of(Literal.ofTimestamp(micros));
```

Comment:
QQ: why there are 2 types of timestamp value from Spark filters?

Reply:
Follow here -> https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala#L86-L87

Reply:
See comment in https://github.com/delta-io/delta/pull/5206#discussion_r2353363769